types/wlr_screencopy_v1.c

```diff
@@ -8,10 +8,85 @@
 #include "wlr-screencopy-unstable-v1-protocol.h"
 #include "util/signal.h"
 
-#define SCREENCOPY_MANAGER_VERSION 1
+#define SCREENCOPY_MANAGER_VERSION 2
+
+struct screencopy_damage {
+	struct wl_list link;
+	struct wlr_output *output;
+	struct pixman_region32 damage;
+	struct wl_listener output_precommit;
+	bool is_handled;
+};
 
 static const struct zwlr_screencopy_frame_v1_interface frame_impl;
 
+static struct screencopy_damage *screencopy_damage_find(
+		struct wlr_screencopy_connection_v1 *conn,
+		struct wlr_output *output) {
+	struct screencopy_damage *damage;
+
+	wl_list_for_each(damage, &conn->damages, link) {
+		if (damage->output == output) {
+			return damage;
+		}
+	}
+
+	return NULL;
+}
+
+static void screencopy_damage_accumulate(struct screencopy_damage *damage) {
+	struct pixman_region32 *region = &damage->damage;
+	struct wlr_output *output = damage->output;
+
+	pixman_region32_union(region, region, &output->pending.damage);
+	pixman_region32_intersect_rect(region, region, 0, 0, output->width,
+		output->height);
+}
+
+static void screencopy_damage_handle_output_precommit(
+		struct wl_listener *listener, void *data) {
+	struct screencopy_damage *damage =
+		wl_container_of(listener, damage, output_precommit);
+	if (!damage->is_handled) {
+		screencopy_damage_accumulate(damage);
+	}
+}
+
+static struct screencopy_damage *screencopy_damage_create(
+		struct wlr_screencopy_connection_v1 *conn,
+		struct wlr_output *output) {
+	struct screencopy_damage *damage =
+		calloc(1, sizeof(struct screencopy_damage));
+	if (!damage) {
+		return NULL;
+	}
+
+	damage->output = output;
+	pixman_region32_init_rect(&damage->damage, 0, 0, output->width,
+		output->height);
+	wl_list_insert(&conn->damages, &damage->link);
+
+	wl_signal_add(&output->events.precommit, &damage->output_precommit);
+	damage->output_precommit.notify =
+		screencopy_damage_handle_output_precommit;
+
+	return damage;
+}
+
+static struct screencopy_damage *screencopy_damage_get(
```

Comment:
Nit: I'd rename this to `screencopy_damage_get_or_create` to make a clear difference with the function that just gets the damage object.